File: fix_accents.js
const fs = require('fs');
const path = require('path');

const FRONTEND_DIR = "c:\\Users\\angel\\Desktop\\QuestEval\\Frontend\\src";
const MOBILE_DIR = "c:\\Users\\angel\\Desktop\\QuestEval\\mobile\\lib";

const wordsToFix = {
    'Evaluaciones': 'Evaluaciones', // fine
    'Evaluacion': 'Evaluación',
    'evaluaciones': 'evaluaciones', // We don't change 'evaluaciones' natively?? Wait, 'evaluacion' -> 'evaluación', 'evaluaciones' -> 'evaluaciones'. IN SPANISH, "evaluaciones" DOES NOT HAVE AN ACCENT! Yes! "evaluación" (singular) has accent, "evaluaciones" (plural) DOES NOT. Oh! This is why! I feel so dumb. "evaluación" -> "evaluaciones". The accent goes away in plural!
    
    'Configuraciones': 'Configuraciones', // No accent in plural
    'Configuracion': 'Configuración',
    'configuracion': 'configuración',
    
    'Exitos': 'Éxitos',
    'exitos': 'éxitos',
    'Exito': 'Éxito',
    'exito': 'éxito',
    
    'Acciones': 'Acciones', // No accent
    'Accion': 'Acción',
    'accion': 'acción',
    
    'Tambien': 'También',
    'tambien': 'también',
    
    'Aqui': 'Aquí',
    'aqui': 'aquí',
    
    'Rubrica': 'Rúbrica',
    'rubrica': 'rúbrica',
    'Rubricas': 'Rúbricas',
    'rubricas': 'rúbricas',
    
    'Calificacion': 'Calificación',
    'Calificaciones': 'Calificaciones', // No accent
    'calificacion': 'calificación',
    
    'Opcion': 'Opción',
    'Opciones': 'Opciones', // No accent
    'opcion': 'opción',
    
    'Version': 'Versión',
    'Versiones': 'Versiones', // No accent
    'version': 'versión',
    
    'Informacion': 'Información',
    'Informaciones': 'Informaciones', // No accent
    'informacion': 'información',
    
    'Descripcion': 'Descripción',
    'Descripciones': 'Descripciones', // No accent
    'descripcion': 'descripción',
    
    'Galeria': 'Galería',
    'galeria': 'galería',
    'Galerias': 'Galerías',
    'galerias': 'galerías', // Yes, retains accent
    
    'Atras': 'Atrás',
    'atras': 'atrás',
    'esito': 'éxito',
};

const riskyLowercase = ['evaluacion', 'configuracion', 'accion', 'rubrica', 'rubricas', 'calificacion', 'opcion', 'version', 'informacion', 'descripcion', 'galeria', 'atras', 'exito', 'galerias', 'exitos'];

function processContent(content, isDart) {
    let newContent = content;

    const replaceCb = (match) => {
        let replacedStr = match;
        for (const [bad, good] of Object.entries(wordsToFix)) {
            if (bad === good) continue;
            
            const regex = new RegExp(`\\b${bad}\\b`, 'g');
            replacedStr = replacedStr.replace(regex, (m, offset, str) => {
                if (!isDart) {
                    let openBrackets = 0;
                    for (let i = 0; i < offset; i++) {
                        if (str[i] === '{') openBrackets++;
                        if (str[i] === '}') openBrackets--;
                    }
                    if (openBrackets > 0) return m; 
                }
                
                if (riskyLowercase.includes(bad)) {
                    // For UI literal patches... prevent changing API paths or variable injections
                    if (match.startsWith("'") || match.startsWith('"') || match.startsWith('\`')) {
                        // Protect against API JSON keys or URLs like '/evaluacion'
                        return m; 
                    }
                }
                
                return good;
            });
        }
        return replacedStr;
    };

    newContent = newContent.replace(/'([^'\\]*(?:\\.[^'\\]*)*)'/g, replaceCb);
    newContent = newContent.replace(/"([^"\\]*(?:\\.[^"\\]*)*)"/g, replaceCb);
    if (!isDart) {
        newContent = newContent.replace(/>([^<]+)</g, replaceCb);
    }

    return newContent;
}

function processFile(filepath) {
    try {
        let content = fs.readFileSync(filepath, 'utf-8');
        let newContent = content;

        if (filepath.endsWith('.tsx') || filepath.endsWith('.ts')) {
            newContent = processContent(content, false);
        } else if (filepath.endsWith('.dart')) {
             newContent = processContent(content, true);
        }

        if (content !== newContent) {
            fs.writeFileSync(filepath, newContent, 'utf-8');
            console.log(`Fixed: ${filepath}`);
        }
    } catch (err) {
        console.error(`Error reading ${filepath}: ${err}`);
    }
}

function walkDir(dir) {
    if (!fs.existsSync(dir)) return;
    const files = fs.readdirSync(dir);
    for (const file of files) {
        const filepath = path.join(dir, file);
        if (filepath.includes('node_modules') || filepath.includes('.next')) continue;
        if (fs.statSync(filepath).isDirectory()) {
            walkDir(filepath);
        } else if (file.endsWith('.ts') || file.endsWith('.tsx') || file.endsWith('.dart')) {
            processFile(filepath);
        }
    }
}

walkDir(FRONTEND_DIR);
walkDir(MOBILE_DIR);

File: fix_accents.py
import os
import re

FRONTEND_DIR = r"c:\Users\angel\Desktop\QuestEval\Frontend\src"
MOBILE_DIR = r"c:\Users\angel\Desktop\QuestEval\mobile\lib"

# Pairs of (Pattern, Replacement)
# Using lookbehind and lookahead to ensure we only replace text that is likely not a variable.
# For lowercase words, we will only replace them if they are preceded by specific Spanish articles or prepositions
# to guarantee they are text, or if they are capitalized (which means they are likely labels).

REPLACEMENTS = [
    # Capitalized (usually labels / UI text)
    (r'\bEvaluacion\b', 'Evaluación'),
    (r'\bConfiguracion\b', 'Configuración'),
    (r'\bExito(?!sa)(?!so)\b', 'Éxito'),
    (r'\bAccion\b', 'Acción'),
    (r'\bTambien\b', 'También'),
    (r'\bAqui\b', 'Aquí'),
    (r'\bRubrica\b', 'Rúbrica'),
    (r'\bRubricas\b', 'Rúbricas'),
    (r'\bCalificacion\b', 'Calificación'),
    (r'\bOpcion\b', 'Opción'),
    (r'\bOpciones\b', 'Opciones'), # Already fine
    (r'\bVersion\b', 'Versión'),
    (r'\bInformacion\b', 'Información'),
    (r'\bDescripcion\b', 'Descripción'),
    (r'\bGaleria\b', 'Galería'),
    (r'\bAtras\b', 'Atrás'),
    (r'\bIniciar sesion\b', 'Iniciar sesión'),
    (r'\bIniciar Sesion\b', 'Iniciar Sesión'),
    (r'\bCerrar sesion\b', 'Cerrar sesión'),
    (r'\bCerrar Sesion\b', 'Cerrar Sesión'),

    # Lowercase but with preceding context to ensure they are text (e.g. "la evaluación")
    (r'\b(la|una|nueva|su|tu|mi|esta|de|con|para|ver|guardar|editar|eliminar|esta|otra) evaluacion\b', r'\1 evaluación'),
    (r'\b(la|una|su|tu|mi|de|con|para) configuracion\b', r'\1 configuración'),
    (r'\b(con) exito\b', r'\1 éxito'),
    (r'\b(la|una|su|tu|mi|de|con|para) accion\b', r'\1 acción'),
    (r'\btambien\b', r'también'), # también is rarely a variable
    (r'\baqui\b', r'aquí'), # aquí is rarely a variable
    (r'\b(la|una|su|tu|mi|de|con|para|nueva|crear|editar|eliminar|esta) rubrica\b', r'\1 rúbrica'),
    (r'\b(las|unas|sus|tus|mis|de|con|para|nuevas|crear|editar|eliminar|estas) rubricas\b', r'\1 rúbricas'),
    (r'\b(la|una|su|tu|mi|de|con|para|nueva|ver|esta) calificacion\b', r'\1 calificación'),
    (r'\b(la|una|su|tu|mi|de|con|para|nueva|esta) opcion\b', r'\1 opción'),
    (r'\b(la|una|su|tu|mi|de|con|para|esta) version\b', r'\1 versión'),
    (r'\b(la|una|su|tu|mi|de|con|para|mas|más) informacion\b', r'\1 información'),
    (r'\b(la|una|su|tu|mi|de|con|para|agregar|editar) descripcion\b', r'\1 descripción'),
    (r'\b(la|una|su|tu|mi|de|con|para|ir a|ver) galeria\b', r'\1 galería'),
    (r'\b(hacia|ir) atras\b', r'\1 atrás'),
    
    # Specific common phrases
    (r'\besito\b', 'éxito'),
    (r'\bexitosa!\b', 'exitosa!'),
    (r'exitosa!', 'exitosa!'), # To trigger if any
]

def process_file(filepath):
    try:
        with open(filepath, 'r', encoding='utf-8') as f:
            content = f.read()
    except Exception as e:
        print(f"Error reading {filepath}: {e}")
        return

    new_content = content
    for pattern, replacement in REPLACEMENTS:
        new_content = re.sub(pattern, replacement, new_content)
        
    # Additional specific UI text replacements that don't match the regex above securely
    new_content = new_content.replace('>evaluacion<', '>evaluación<')
    new_content = new_content.replace('>Evaluaciones<', '>Evaluaciones<') # Fine
    new_content = new_content.replace("'evaluacion'", "'evaluación'") # Be careful if evaluating map keys, but usually it's text. Let's skip replacing exact enclosed quotes to avoid breaking JSON keys.
    new_content = new_content.replace('>Rubrica<', '>Rúbrica<')
    new_content = new_content.replace('>Calificacion<', '>Calificación<')
    new_content = new_content.replace('>Descripcion<', '>Descripción<')
    new_content = new_content.replace('>Galeria<', '>Galería<')

    if new_content != content:
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(new_content)
        print(f"Fixed accents in: {filepath}")

def main():
    for root_dir in [FRONTEND_DIR, MOBILE_DIR]:
        for dirpath, dirnames, filenames in os.walk(root_dir):
            for filename in filenames:
                if filename.endswith(('.ts', '.tsx', '.dart')):
                    process_file(os.path.join(dirpath, filename))

if __name__ == "__main__":
    main()
